### PR TITLE
Reduce duplicated code in Analytics classes

### PR DESCRIPTION
**Pull request by tomalec**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This is a feature branch to gather multiple improvements to the Analytics codebase with the goal to:
- Make it easier for extension developers to reuse & extend
- Make it easier for our Team to maintain

- https://github.com/woocommerce/woocommerce/pull/48964
- https://github.com/woocommerce/woocommerce/pull/49320
- https://github.com/woocommerce/woocommerce/pull/49427
- https://github.com/woocommerce/woocommerce/pull/50746
- https://github.com/woocommerce/woocommerce/pull/50748
- https://github.com/woocommerce/woocommerce/pull/50749
- https://github.com/woocommerce/woocommerce/pull/50799

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

All PRs were reviewed separately, but to double-check

1. Smoke test analytics for regression
2. Review the code once again

Apart from PRs, I pushed one commit to set the target release version in `since`tags https://github.com/woocommerce/woocommerce/pull/49425/commits/e3286534e6e96135858e2b0712c1dfce6851279c

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by tomalec** [comment]
Squashed commit message:
```
- Add code docs to the Analytics classes and make get_order_statuses non-internal.
   Document analytics classes and some methods. Add simple examples of how to use it. Give a brief "why?" and "when?" answers. Add links between related classes.
- Replace `Automattic\WooCommerce\Admin\API\Reports\*\Query` classes with a single `GenericQuery` class.
- Reduce the amount of duplicated code in Analytics `DataStore`s.
- Reduce duplicated code in Analytics controllers
- Add `FilteredGetDataTrait`, `OrderAwareControllerTrait`, and `StatsDataStoreTrait` for extension developers to reuse while creating custom Analytics
- Add a `GenericQuery` to reduce duplicated code in Query classes.
   Also, to expose it to community extensions for reuse.
- Use `GenericQuery` instead of duplicated `Query` classes
- Move caching code to shared `DataStore::get_data`
- Move intervals specific `initialize_queries` to shared Trait
- Move intervals checking logic to shared `StatsDataStoreTrait`
- Reuse `GenericController::prepare_item_for_response` for `Reports\Controller` subclasses
- Reuse `GenericController::get_collection_params` for `Reports\Controller` subclasses.
- Move `get_items` code to shared `GenericStatsController`
- Move shared paginable controllers `get_items` code to the generic class
- Move fields param to `GenericStatsController`
- Separate `OrderAwareControllerTrait` from `ReportController`
   to allow specific Analytics Controllers to extend the `Generic(Stats)Controller` directly, without extending the `ReportController`, which is used to list reports.


It's meant not to change any behavior. However, due to unification, it did tweak a few things mostly from the developer perspective:
- Previously, only some controllers threw an error from `get_items` when the data was missing; some did not. Now controllers the following Controllers changed behavior to also for such an error:
	- Coupons
	- Customers
	- Downloads
	- Orders
	- Products
	- Taxes
- the error has one error code from all controllers:`woocommerce_rest_reports_invalid_response` (previously there were report-specific)
   - `woocommerce_rest_reports_categories_invalid_response` -> `woocommerce_rest_reports_invalid_response`
- In Orders Controller, the `$orders_data['order_number'], $orders_data['total_formatted']` are set in the `prepare_item_for_response` method, not in the `get_items`So `prepare_item_for_response` function is expected to be called with bare data from Query
- In Products Controller extended-info is now sanitized in the `prepare_item_for_response` method, not in the `get_items`
- ⚠️ (breaking) Previosly some Controlers' `prepare_item_for_response` function was getting and expecting first argument to be casted to `object`, while the rest of controllers used data as returned from the Store. This PR unifies this behavior  for
   -  `Coupons\Stats\Controller`
   -  `Taxes\Controller`
   -  `Taxes\Stats\Controller`
   So, if you extend those classes and override `prepare_item_for_response` in a way that expects `object`, you will get an error. (I have not found such cases in the `all-plugins` repo.)
   All `woocommerce_rest_prepare_report_*` filters remain intact and inconsistent for backward compatibility.
```


I'll also add another changelog entry to reflect the above

File: plugins/woocommerce/src/Admin/API/Reports/Categories/Query.php [hunk 1]
```diff
@@ -21,7 +21,9 @@
 use Automattic\WooCommerce\Admin\API\Reports\Query as ReportsQuery;
 
 /**
- * API\Reports\Query
+ * API\Reports\Categories\Query
+ *
+ * @deprecated x.x.x Categories\Query class is deprecated, please use GenericQuery or \WC_Object_Query instead.
```
Comment (ianlin) [inline plugins/woocommerce/src/Admin/API/Reports/Categories/Query.php]: There are still some parts that have `x.x.x`, do we need to also update them to `9.3.0`?

**Review by ianlin — COMMENTED** [review]
I have ran through some smoke tests locally, tldr: all good.

- ✅ All the core analytics worked normally
- ✅ The [internal example plugin](https://github.com/woocommerce/current-wc-analytics-examples) worked normally
  - I need to change the name of data store to the `rest_base`, but it's because it was built from the previous developing version, and I didn't implement the new method `get_datastore_data()`.
- ✅ Installed some plugins that have WC Analytics integrated, all good.
  - WooCommerce Product Bundle
  - Gift Cards
  - WooCommerce Coupon Campaigns
  - AutomateWoo.

For the latest stable AutomateWoo `6.0.32`, I found there was a fatal, but it was already [fixed in here](https://github.com/woocommerce/automatewoo/pull/1838/files#diff-b563c71eabb821344b4abcfec74a60417e68fd990564034c00d78f1aae276afcL116-R118), and the newer version of AW will be released earlier than WC `9.3`. Confirmed that using AW from `develop` branch and the fatal was gone. I also tried to use the combination of AW from `develop` branch and the latest WC `9.1.4`, all worked as expected.

**Comment by tomalec** [comment]
I also run tests with AutomateWoo Referrals with this change https://github.com/woocommerce/automatewoo-referrals/pull/310 and the `trunk` version of WC. Looks good.

**Review by ianlin — APPROVED** [review]
Like the previous review everything worked normally with this refactoring. LGTM.

**Pull request merged**